File: packages/reference-engine/src/cel.js
'use strict';
/**
 * CEL 表达式求值 —— 参考实现只覆盖内置资产实际用到的子集。
 *
 * 真正的引擎应使用标准 CEL 实现(cel-java),本文件的作用是让参考引擎
 * 能跑通那 3 条使用 CEL 的策略,并验证 packages/cel-functions/ 中定义的
 * 函数语义(尤其是边界行为)是否可实现、是否有歧义。
 *
 * 支持的子集:
 *   inTimeWindow("HH:MM", "HH:MM")
 *   ipLocation(field, "level") == "值"  /  != /  in [...]
 *   checkNotice(keyType, keyValue, strategyName, withinSeconds) > N
 *   !( ... )
 */

const DEFAULT_TZ_OFFSET_MIN = 8 * 60;    // Asia/Shanghai,对应 nebula.timezone 默认值

let locationResolver = () => 'unknown';  // 规格:查询失败返回 "unknown"
function setLocationResolver(fn) { locationResolver = fn; }

let tzOffsetMin = DEFAULT_TZ_OFFSET_MIN;
function setTimezoneOffsetMinutes(min) { tzOffsetMin = min; }

/** 事件时间在部署时区下的「当日分钟数」 */
function minutesOfDay(tsMs) {
  const local = tsMs + tzOffsetMin * 60 * 1000;
  const dayMs = ((local % 86400000) + 86400000) % 86400000;
  return Math.floor(dayMs / 60000);
}

function parseHHMM(s) {
  const m = /^(\d{1,2}):(\d{2})$/.exec(String(s).trim());
  if (!m) throw new Error(`inTimeWindow 的时刻格式应为 HH:MM,实际: ${s}`);
  return parseInt(m[1], 10) * 60 + parseInt(m[2], 10);
}

/**
 * keyType 归一化。
 *
 * 规格最初写的取值是 `ip` / `uid` / `did` / `page` —— 那是 1.x 的词汇,而 2.0 的名单
 * 主体类型是 check_type(IP / USER / DeviceID / OrderID)。**`page` 没有对应的名单
 * 类型**,也就是说按它查永远查不到东西。
 *
 * 这里接受 check_type 本身,也接受 1.x 的三个别名。`page` 与其它未知取值一律抛错 ——
 * 静默返回 0 会让一条永远不命中的策略看起来在正常工作。
 */
const KEY_TYPES = { ip: 'IP', uid: 'USER', did: 'DeviceID', order_id: 'OrderID' };
const CHECK_TYPES = new Set(['IP', 'USER', 'DeviceID', 'OrderID']);

function normalizeKeyType(t) {
  const s = String(t);
  if (CHECK_TYPES.has(s)) return s;
  const alias = KEY_TYPES[s.toLowerCase()];
  if (alias) return alias;
  throw new Error(
    `checkNotice 的 keyType 取值非法: ${t}(可取 ${[...CHECK_TYPES].join(' / ')},`
    + `或 1.x 别名 ip / uid / did / order_id)`);
}

const FUNCTIONS = {
  /** 规格:start 含、end 不含;start > end 表示跨零点;按部署时区判断 */
  inTimeWindow(args, event) {
    const start = parseHHMM(args[0]);
    const end = parseHHMM(args[1]);
    const now = minutesOfDay(event.timestamp);
    if (start <= end) return now >= start && now < end;
    return now >= start || now < end;          // 跨零点
  },

  /**
   * 策略级联:该主体在过去 withinSeconds 内命中 strategyName 的次数。
   *
   * 规格:1.x SPL 里唯一真正可用的业务函数($CHECKNOTICE)。
   *
   * **数的是已产出的告警,不含被去重压掉的那些。** 去重意味着「这条告警没有被报出去」,
   * 而级联判定的语义是「之前报过没有」—— 把压掉的也算进来会让下游看到一个它从未收到过
   * 的依据。这条规格里没写,是实现时定的,写在这里与 Java 端保持一致。
   *
   * 时间窗是 **[now - withinSeconds*1000, now)** —— 起点含,终点不含。终点不含是
   * 关键:同一条事件里先求值的策略可能刚产出一条告警,把它算进来会让结果依赖策略的
   * 求值顺序,而顺序不是契约。checkNotice 看到的是「在这条事件之前已经报过的」。
   */
  checkNotice(args, event, ctx) {
    const [keyType, keyValue, strategyName, withinSeconds] = args;
    const type = normalizeKeyType(keyType);
    if (!ctx || typeof ctx.countNotices !== 'function') {
      throw new Error('checkNotice 需要告警历史,当前求值上下文没有提供');
    }
    const within = Number(withinSeconds);
    if (!Number.isFinite(within) || within <= 0) {
      throw new Error(`checkNotice 的 withinSeconds 应为正整数,实际: ${withinSeconds}`);
    }
    return ctx.countNotices(type, String(keyValue ?? ''), String(strategyName),
      event.timestamp - within * 1000, event.timestamp);
  },

  /** 规格:查询失败或无结果返回 "unknown",不抛异常 */
  ipLocation(args) {
    const ip = args[0];
    const level = args[1];
    if (ip === null || ip === undefined || ip === '') return 'unknown';
    try {
      const r = locationResolver(ip, level);
      return r === null || r === undefined || r === '' ? 'unknown' : r;
    } catch (_) {
      return 'unknown';
    }
  },
};

/** 极简求值器:够用即可,不追求完整 CEL 语法 */
function evalCel(expr, event, ctx) {
  const src = String(expr).trim();

  // !( ... )
  if (src.startsWith('!(') && src.endsWith(')')) {
    return !evalCel(src.slice(2, -1), event, ctx);
  }

  // <call> in [ ... ]
  const inMatch = /^(.+?)\s+in\s+(\[.*\])$/.exec(src);
  if (inMatch) {
    const left = evalTerm(inMatch[1], event, ctx);
    const list = JSON.parse(inMatch[2]);
    return list.map(String).includes(String(left));
  }

  // <call> > N / >= N / < N / <= N —— checkNotice 返回整数,需要数值比较
  const num = /^(.+?)\s*(>=|<=|>|<)\s*(-?\d+(?:\.\d+)?)$/.exec(src);
  if (num) {
    const left = Number(evalTerm(num[1], event, ctx));
    const right = Number(num[3]);
    switch (num[2]) {
      case '>': return left > right;
      case '<': return left < right;
      case '>=': return left >= right;
      default: return left <= right;
    }
  }

  // <call> == "值"  /  != "值"
  const cmp = /^(.+?)\s*(==|!=)\s*(.+)$/.exec(src);
  if (cmp) {
    const left = evalTerm(cmp[1], event, ctx);
    const right = stripQuotes(cmp[3].trim());
    return cmp[2] === '==' ? String(left) === right : String(left) !== right;
  }

  // 单独的布尔函数调用
  const v = evalTerm(src, event, ctx);
  return !!v;
}

function evalTerm(src, event, ctx) {
  const s = src.trim();
  const call = /^([A-Za-z_][A-Za-z0-9_]*)\((.*)\)$/.exec(s);
  if (call) {
    const name = call[1];
    const fn = FUNCTIONS[name];
    if (!fn) {
      throw new Error(`未实现的 CEL 函数: ${name}(定义见 packages/cel-functions/README.md)`);
    }
    const args = splitArgs(call[2]).map((a) => {
      const t = a.trim();
      if (/^".*"$/.test(t) || /^'.*'$/.test(t)) return stripQuotes(t);
      if (/^-?\d+(\.\d+)?$/.test(t)) return Number(t);
      return event[t];                    // 裸标识符 = 事件字段
    });
    return fn(args, event, ctx);
  }
  if (/^".*"$/.test(s) || /^'.*'$/.test(s)) return stripQuotes(s);
  return event[s];
}

function splitArgs(s) {
  const out = []; let depth = 0; let cur = ''; let quote = null;
  for (const ch of s) {
    if (quote) { cur += ch; if (ch === quote) quote = null; continue; }
    if (ch === '"' || ch === "'") { quote = ch; cur += ch; continue; }
    if (ch === '(' || ch === '[') depth++;
    if (ch === ')' || ch === ']') depth--;
    if (ch === ',' && depth === 0) { out.push(cur); cur = ''; continue; }
    cur += ch;
  }
  if (cur.trim() !== '') out.push(cur);
  return out;
}

function stripQuotes(s) {
  const t = String(s).trim();
  if ((t.startsWith('"') && t.endsWith('"')) || (t.startsWith("'") && t.endsWith("'"))) {
    return t.slice(1, -1);
  }
  return t;
}

module.exports = { evalCel, setLocationResolver, setTimezoneOffsetMinutes, minutesOfDay };

File: apps/engine/src/main/java/cn/threathunter/nebula/engine/rule/Cel.java
package cn.threathunter.nebula.engine.rule;

import java.time.Instant;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * CEL 表达式求值 —— 只覆盖内置资产实际用到的子集。
 *
 * <p>生产引擎最终应改用标准 cel-java;本实现的作用是让引擎能跑通那 3 条使用 CEL
 * 的策略,并与参考实现的 {@code src/cel.js} 逐条对齐。函数语义的规范定义见
 * {@code packages/cel-functions/README.md}。
 *
 * <p>支持:{@code inTimeWindow("HH:MM","HH:MM")}、
 * {@code checkNotice(keyType,keyValue,strategyName,withinSeconds) > N}、
 * {@code ipLocation(field,"level") == "值"}、{@code ... in [...]}、{@code !(...)}。
 */
public final class Cel {

    private Cel() {
    }

    /** 部署时区。规格:「深夜」是业务含义,必须按当地时间判断,不是 UTC。 */
    private static volatile ZoneId zone = ZoneId.of("Asia/Shanghai");

    public static void setZone(ZoneId z) {
        zone = z == null ? ZoneId.of("Asia/Shanghai") : z;
    }

    @FunctionalInterface
    public interface LocationResolver {
        String resolve(String ip, String level);
    }

    private static volatile LocationResolver locationResolver = (ip, level) -> "unknown";

    public static void setLocationResolver(LocationResolver r) {
        locationResolver = r == null ? (ip, level) -> "unknown" : r;
    }

    private static final Pattern CALL = Pattern.compile("^([A-Za-z_][A-Za-z0-9_]*)\\((.*)\\)$");
    private static final Pattern IN = Pattern.compile("^(.+?)\\s+in\\s+(\\[.*])$");
    private static final Pattern CMP = Pattern.compile("^(.+?)\\s*(==|!=)\\s*(.+)$");
    /** checkNotice 返回整数,需要数值比较。 */
    private static final Pattern NUM =
            Pattern.compile("^(.+?)\\s*(>=|<=|>|<)\\s*(-?\\d+(?:\\.\\d+)?)$");

    /**
     * 告警历史查询,供 {@code checkNotice} 做策略级联判定。
     *
     * <p>区间两端都含。{@code null} 表示当前求值上下文没有历史可查 —— 那时调用
     * checkNotice 会抛,而不是返回 0:返回 0 会让一条永远不命中的策略看起来在正常工作。
     */
    @FunctionalInterface
    public interface NoticeHistory {
        int count(String checkType, String key, String strategyName, long fromTs, long toTs);
    }

    /**
     * keyType 归一化。
     *
     * <p>规格最初写的取值是 {@code ip} / {@code uid} / {@code did} / {@code page} ——
     * 那是 1.x 的词汇,而 2.0 的名单主体类型是 check_type。<b>{@code page} 没有对应的
     * 名单类型</b>,按它查永远查不到东西,所以不接受它。
     */
    private static String normalizeKeyType(String t) {
        switch (t) {
            case "IP", "USER", "DeviceID", "OrderID" -> {
                return t;
            }
            default -> {
                String alias = switch (t.toLowerCase()) {
                    case "ip" -> "IP";
                    case "uid" -> "USER";
                    case "did" -> "DeviceID";
                    case "order_id" -> "OrderID";
                    default -> null;
                };
                if (alias == null) {
                    throw new IllegalArgumentException(
                            "checkNotice 的 keyType 取值非法: " + t
                            + "(可取 IP / USER / DeviceID / OrderID,或 1.x 别名 ip / uid / did / order_id)");
                }
                return alias;
            }
        }
    }
    private static final Pattern HHMM = Pattern.compile("^(\\d{1,2}):(\\d{2})$");

    public static boolean eval(String expr, Map<String, Object> event) {
        return eval(expr, event, null);
    }

    public static boolean eval(String expr, Map<String, Object> event, NoticeHistory history) {
        String s = expr == null ? "" : expr.trim();
        if (s.startsWith("!(") && s.endsWith(")")) {
            return !eval(s.substring(2, s.length() - 1), event, history);
        }
        Matcher in = IN.matcher(s);
        if (in.matches()) {
            Object left = term(in.group(1), event, history);
            for (String item : parseList(in.group(2))) {
                if (item.equals(String.valueOf(left))) {
                    return true;
                }
            }
            return false;
        }
        Matcher num = NUM.matcher(s);
        if (num.matches()) {
            double left = toDouble(term(num.group(1), event, history));
            double right = Double.parseDouble(num.group(3));
            return switch (num.group(2)) {
                case ">" -> left > right;
                case "<" -> left < right;
                case ">=" -> left >= right;
                default -> left <= right;
            };
        }
        Matcher cmp = CMP.matcher(s);
        if (cmp.matches()) {
            Object left = term(cmp.group(1), event, history);
            String right = stripQuotes(cmp.group(3).trim());
            boolean eq = String.valueOf(left).equals(right);
            return "==".equals(cmp.group(2)) == eq;
        }
        Object v = term(s, event, history);
        return Boolean.TRUE.equals(v);
    }

    private static double toDouble(Object v) {
        if (v instanceof Number n) {
            return n.doubleValue();
        }
        try {
            return Double.parseDouble(String.valueOf(v));
        } catch (NumberFormatException e) {
            return Double.NaN;   // NaN 与任何数比较都为假 —— 与「取不到值就不命中」一致
        }
    }

    private static Object term(String src, Map<String, Object> event, NoticeHistory history) {
        String s = src.trim();
        Matcher m = CALL.matcher(s);
        if (m.matches()) {
            String fn = m.group(1);
            List<Object> args = new ArrayList<>();
            for (String a : splitArgs(m.group(2))) {
                String t = a.trim();
                if (t.startsWith("\"") || t.startsWith("'")) {
                    args.add(stripQuotes(t));
                } else if (t.matches("-?\\d+(\\.\\d+)?")) {
                    args.add(Double.parseDouble(t));
                } else {
                    args.add(event.get(t)); // 裸标识符 = 事件字段
                }
            }
            return call(fn, args, event, history);
        }
        if (s.startsWith("\"") || s.startsWith("'")) {
            return stripQuotes(s);
        }
        return event.get(s);
    }

    private static Object call(String fn, List<Object> args, Map<String, Object> event,
                              NoticeHistory history) {
        switch (fn) {
            case "checkNotice": {
                // 语义与参考引擎逐条对齐,见 packages/reference-engine/src/cel.js。
                // 数的是**已产出**的告警,不含被去重压掉的那些 —— 去重意味着这条告警
                // 没有被报出去,而级联判定问的是「之前报过没有」。
                if (history == null) {
                    throw new IllegalStateException(
                            "checkNotice 需要告警历史,当前求值上下文没有提供");
                }
                String type = normalizeKeyType(String.valueOf(args.get(0)));
                String key = args.get(1) == null ? "" : String.valueOf(args.get(1));
                String strategy = String.valueOf(args.get(2));
                double within = toDouble(args.get(3));
                if (!(within > 0)) {
                    throw new IllegalArgumentException(
                            "checkNotice 的 withinSeconds 应为正整数,实际: " + args.get(3));
                }
                Object ts = event.get("timestamp");
                if (!(ts instanceof Number n)) {
                    return 0;
                }
                long now = n.longValue();
                return history.count(type, key, strategy, now - (long) (within * 1000), now);
            }
            case "inTimeWindow": {
                int start = parseHHMM(String.valueOf(args.get(0)));
                int end = parseHHMM(String.valueOf(args.get(1)));
                Object ts = event.get("timestamp");
                if (!(ts instanceof Number n)) {
                    return false;
                }
                ZonedDateTime t = Instant.ofEpochMilli(n.longValue()).atZone(zone);
                int now = t.getHour() * 60 + t.getMinute();
                // 规格:start 含、end 不含;start > end 表示跨零点
                return start <= end ? (now >= start && now < end) : (now >= start || now < end);
            }
            case "ipLocation": {
                Object ip = args.get(0);
                if (ip == null || String.valueOf(ip).isEmpty()) {
                    return "unknown";
                }
                String level = args.size() > 1 ? String.valueOf(args.get(1)) : "province";
                try {
                    String r = locationResolver.resolve(String.valueOf(ip), level);
                    // 规格:查询失败或无结果返回 unknown,不抛异常
                    return r == null || r.isEmpty() ? "unknown" : r;
                } catch (RuntimeException e) {
                    return "unknown";
                }
            }
            default:
                throw new IllegalArgumentException(
                        "未实现的 CEL 函数: " + fn + "(定义见 packages/cel-functions/README.md)");
        }
    }

    private static int parseHHMM(String s) {
        Matcher m = HHMM.matcher(s.trim());
        if (!m.matches()) {
            throw new IllegalArgumentException("inTimeWindow 的时刻格式应为 HH:MM,实际: " + s);
        }
        return Integer.parseInt(m.group(1)) * 60 + Integer.parseInt(m.group(2));
    }

    private static List<String> parseList(String s) {
        List<String> out = new ArrayList<>();
        String body = s.trim();
        if (body.startsWith("[")) {
            body = body.substring(1);
        }
        if (body.endsWith("]")) {
            body = body.substring(0, body.length() - 1);
        }
        for (String p : splitArgs(body)) {
            String t = p.trim();
            if (!t.isEmpty()) {
                out.add(stripQuotes(t));
            }
        }
        return out;
    }

    private static List<String> splitArgs(String s) {
        List<String> out = new ArrayList<>();
        int depth = 0;
        char quote = 0;
        StringBuilder cur = new StringBuilder();
        for (char ch : s.toCharArray()) {
            if (quote != 0) {
                cur.append(ch);
                if (ch == quote) {
                    quote = 0;
                }
                continue;
            }
            if (ch == '"' || ch == '\'') {
                quote = ch;
                cur.append(ch);
                continue;
            }
            if (ch == '(' || ch == '[') {
                depth++;
            }
            if (ch == ')' || ch == ']') {
                depth--;
            }
            if (ch == ',' && depth == 0) {
                out.add(cur.toString());
                cur.setLength(0);
                continue;
            }
            cur.append(ch);
        }
        if (!cur.toString().isBlank()) {
            out.add(cur.toString());
        }
        return out;
    }

    private static String stripQuotes(String s) {
        String t = s.trim();
        if (t.length() >= 2 && ((t.startsWith("\"") && t.endsWith("\""))
                || (t.startsWith("'") && t.endsWith("'")))) {
            return t.substring(1, t.length() - 1);
        }
        return t;
    }
}
